plugins/fsverity.c

```diff
@@ -0,0 +1,168 @@
+    int rc = RPMRC_OK;
+    int fd;
+    rpmFileAction action = XFO_ACTION(op);
+    char *buffer;
+
+    /* Ignore skipped files and unowned directories */
+    if (XFA_SKIPPING(action) || (op & FAF_UNOWNED)) {
+	rpmlog(RPMLOG_DEBUG, "fsverity skipping early: path %s dest %s\n",
+	       path, dest);
+	goto exit;
+    }
+
+    /*
+     * Do not install signatures for config files unless the
+     * user explicitly asks for it.
+     */
+    if (rpmfiFFlags(fi) & RPMFILE_CONFIG) {
+	if (!(rpmfiFMode(fi) & (S_IXUSR|S_IXGRP|S_IXOTH)) &&
+	    !sign_config_files) {
+	    rpmlog(RPMLOG_DEBUG, "fsverity skipping: path %s dest %s\n",
+		   path, dest);
+
+	    goto exit;
+	}
+    }
+
+    /*
+     * Right now fsverity doesn't deal with symlinks or directories, so do
+     * not try to install signatures for non regular files.
+     */
+    if (!S_ISREG(rpmfiFMode(fi))) {
+	rpmlog(RPMLOG_DEBUG, "fsverity skipping non regular: path %s dest %s\n",
+	       path, dest);
+	goto exit;
+    }
+
+    signature = rpmfiVSignature(fi, &len, &algo, &blksz);
+    if (!signature || !len) {
+	rpmlog(RPMLOG_DEBUG, "fsverity no signature for: path %s dest %s\n",
+	       path, dest);
+	goto exit;
+    }
+
+    memset(&arg, 0, sizeof(arg));
+    arg.version = 1;
+    if (algo)
+	arg.hash_algorithm = algo;
+    else
+	arg.hash_algorithm = FS_VERITY_HASH_ALG_SHA256;
+    if (blksz)
+	arg.block_size = blksz;
+    else
+	arg.block_size = sysconf(_SC_PAGESIZE);
+    arg.sig_ptr = (uintptr_t)signature;
+    arg.sig_size = len;
+
+    buffer = pgpHexStr(signature, arg.sig_size);
+    rpmlog(RPMLOG_DEBUG, "applying signature: %s\n", buffer);
+    free(buffer);
+
+    fd = open(path, O_RDONLY);
+    if (fd < 0) {
+	rpmlog(RPMLOG_ERR, "failed to open path %s\n", path);
+	goto exit;
+    }
+
+    if (ioctl(fd, FS_IOC_ENABLE_VERITY, &arg) != 0) {
+	switch(errno) {
+	case EBADMSG:
+	    rpmlog(RPMLOG_DEBUG, "invalid fsverity signature for %s\n", path);
+	    break;
+	case EEXIST:
+	    rpmlog(RPMLOG_DEBUG, "fsverity signature already enabled %s\n",
+		   path);
+	    break;
+	case EINVAL:
+	    rpmlog(RPMLOG_DEBUG, "invalid arguments for ioctl %s\n", path);
+	    break;
+	case EKEYREJECTED:
+	    rpmlog(RPMLOG_DEBUG, "signature doesn't match file %s\n", path);
+	    break;
+	case EMSGSIZE:
+	    rpmlog(RPMLOG_DEBUG, "invalid signature size for %s\n", path);
+	    break;
+	case ENOPKG:
+	    rpmlog(RPMLOG_DEBUG, "unsupposed signature algoritm (%i) for %s\n",
+		   arg.hash_algorithm, path);
+	    break;
+	case ENOTTY:
+	    rpmlog(RPMLOG_DEBUG, "fsverity not supported by file system for %s\n",
+		   path);
+	    break;
+	case EOPNOTSUPP:
+	    rpmlog(RPMLOG_DEBUG, "fsverity not enabled on file system for %s\n",
+		   path);
+	    break;
+	case ETXTBSY:
+	    rpmlog(RPMLOG_DEBUG, "file is open by other process %s\n",
+		   path);
+	    break;
+	default:
+	    rpmlog(RPMLOG_DEBUG, "failed to enable verity (errno %i) for %s\n",
+		   errno, path);
+	    break;
+	}
```

Comment:
AFAICT these failures should result in RPMRC_FAIL return code as the plugin failed to accomplish its task, accompanied by an actual error message. I'd suggest something like this to replace the big switch:
```
if (ioctl() ... != 0) {
    rpmlog(RPMLOG_ERR, _("enabling verity failed: %s %s\n", path), strerror(errno));
    rc = RPMRC_FAIL;
}
```

Reply:
I did this on purpose as I didn't want the installation to fail.

fsverity support is on a per filesystem level, and I deliberately ignore the failures silently on file systems that do not have the support.

If someone builds all their rpms with verity signatures, but not all their systems have fsverity enabled/supported in the file system, I don't think they want to get a large number of errors when trying to install packages.

There is also the situation where someone has a setup where one partition has fsverity enabled and one where it isn't. If parts of the rpm are installed on both partitions, this could lead to some tricky failure situations.

Reply:
Ignoring on unsupported filesystems is totally fine, that's exactly what we do for SELinux (which is not supported on all filesystems either). There might well be other specific errors where ignoring is the right thing to do, but in general errors should be treated as such and fail the install, or at least issue a warning that something might not be okay.

Reply:
> Ignoring on unsupported filesystems is totally fine, that's exactly what we do for SELinux (which is not supported on all filesystems either). There might well be other specific errors where ignoring is the right thing to do, but in general errors should be treated as such and fail the install, or at least issue a warning that something might not be okay.

OK, excuse my ignorance here. If I return RPMRC_FAIL for these, will it fail the installation of the package, or just spew warnings? I didn't want to get into a state where I had to check the file system for each file, before I enabled verity on the file.

Do you prefer I return RPMRC_OK if it's not supported, but RPMRC_FAIL if it actually fails? I'll have to have another look at all the return values to make sure I have them covered correctly.

Reply:
Returning FAIL here will fail the package install (which isn't pretty, but sometimes necessary). The exact semantics of when its okay to ignore failure are case-dependent and subject to debate and often practical experiences too. I think it's unreasonable to expect getting this 100% right from the go, consider this feedback more of a guideline than specific demands: silencing and/or ignoring an error should always be an exception to the rule.

Reply:
I went through this and changed it to return RPMRC_FAIL for all cases, except for the error cases where fsverity is not enabled or supported by the file system. I think it's fair to fail hard and ugly for the cases where the signature is malformed etc.